### ISSUE TITLE
LSTM loss and accuracy values differ greatly between GPU and CPU implementation

### ISSUE THREAD

**Issue by JoshJansenVanVuren** [tensorflow/tensorflow#40805]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from: conda
- TensorFlow version: 2.2.0
- Python version: 3.7.7
- CUDA/cuDNN version: CUDA Version 10.1.243
- GPU model and memory: Nvidia GeForce 920M

**Describe the current behavior**
I've attached a <code>main.py</code> that should be able to replicate the error. When training a LSTM model the results differ by an order of magnitude (in the log sense) between GPU and CPU implementations.

(I've also attached the <code>dataset.txt</code> file that I quickly parsed because I cannot share the corpus
I am currently utilizing.)

I've tested over multiple runs and this effect doesn't occur on the GRU RNN, only LSTM.

Its rather hacky but I've used
<code> os.environ['CUDA_VISIBLE_DEVICES'] = '-1' </code>
to switch between GPU and CPU implementations and there is on <code>line 12</code> in <code>main.py</code>

GPU:
    Epoch 1 Train Loss 7.8753 Train Accuracy 46.6 Time 17.16
    Epoch 2 Train Loss 7.1481 Train Accuracy 49.0 Time 11.74 
    Epoch 3 Train Loss 6.9079 Train Accuracy 49.0 Time 13.47 

CPU:
    Epoch 1 Train Loss 6.6438 Train Accuracy 39.8 Time 16.51
    Epoch 2 Train Loss 4.6425 Train Accuracy 40.4 Time 11.57
    Epoch 3 Train Loss 4.4195 Train Accuracy 40.4 Time 11.73

**Describe the expected behavior**
I understand there should be small floating point error difference, as well as difference between random initialization, but the
loss values should at least be similar.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem.

Apologies, this may not be exact minimum:

[dataset.txt](https://github.com/tensorflow/tensorflow/files/4833424/dataset.txt)

```
import os
import sys
import time
from datetime import datetime
import tensorflow as tf
import tensorflow_datasets as tfds
import math
import random
import numpy as np

# the important bit 
#os.environ['CUDA_VISIBLE_DEVICES'] = '-1'

class LSTM_generator(tf.keras.Model):
    def __init__(self, vocab_size, embedding_dim, hidden_units, batch_sz):
        super(LSTM_generator, self).__init__()
        self.batch_sz = batch_sz
        self.hidden_units = hidden_units
        self.embedding = tf.keras.layers.Embedding(input_dim=vocab_size,
                                                   output_dim=embedding_dim,
                                                   mask_zero=True)
        self.LSTM = tf.keras.layers.LSTM(self.hidden_units,
                                        return_sequences=True,
                                        return_state=True)
        self.fc = tf.keras.layers.Dense(vocab_size)

    def call(self, inp, states):
        mask = self.embedding.compute_mask(inp)
        
        x = self.embedding(inp)

        output, hidden, cell = self.LSTM(x,initial_state=states,mask=mask)

        x = self.fc(output)

        return x, (hidden, cell)

    def initialize_hidden_state(self):
        return tf.zeros((self.batch_sz,self.hidden_units)), tf.zeros((self.batch_sz,self.hidden_units))

# Converts str to list of str word 'tokens'

def tokenize(inp):
    words = []
    buffer = ""
    for let in inp:
        if chr(let) == " ":
            words.append(buffer)
            buffer = ""
        else:
            buffer = buffer + chr(let)

    words.append(buffer)

    return words

def split_input_target(sequence):   
    input_text = sequence[:-1]
    target_text = sequence[1:]
    return input_text, target_text

dataset_path = 'dataset.txt'
aug_dataset_path = 'aug.txt'
vocabulary_set = set()

dataset = tf.data.TextLineDataset(dataset_path)

# get vocab
for text_tensor in dataset:
    some_tokens = tokenize(text_tensor.numpy())
    vocabulary_set.update(some_tokens)

encoder = tfds.features.text.TokenTextEncoder(vocabulary_set)

def encode(text_tensor):
    encoded_text = encoder.encode(text_tensor.numpy())

    return tf.cast(encoded_text,tf.int64)

def encode_map_fn(text):
    encoded_text = tf.py_function(encode,inp=[text],Tout=tf.int64)

    encoded_text.set_shape([None])

    return encoded_text

raw_dataset = dataset.map(encode_map_fn)

# split dataset into sources and targets
dataset_split = raw_dataset.map(split_input_target)

# model parameters 
embedding_dim = 8
rnn_units = 64
batch_size = 64
epochs = 60
sens_to_generate = 8000
temperature = 1.0
vocab_size = encoder.vocab_size
max_seq_len_to_generate = 30
BUFFER_SIZE = 10000

# pad and batch the training datasets
dataset_batched_padded = dataset_split.padded_batch(batch_size,
                                        padded_shapes=([None,],[None,]),
                                        drop_remainder=True).shuffle(BUFFER_SIZE)

generator = LSTM_generator(vocab_size, embedding_dim, rnn_units, batch_size)

# signatures help tf run faster - tracing or something
train_step_signature = [
    tf.TensorSpec(shape=(batch_size, None), dtype=tf.int64),
    tf.TensorSpec(shape=(batch_size, None), dtype=tf.int64),
    (tf.TensorSpec(shape=(batch_size, rnn_units), dtype=tf.float32),
    tf.TensorSpec(shape=(batch_size, rnn_units), dtype=tf.float32))]

optimizer = tf.keras.optimizers.Adam()

loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
                        from_logits=True, reduction='none')           

def loss_function(real, pred):
    return tf.reduce_mean(loss_object(real, pred))

# Define metrics
train_loss = tf.keras.metrics.Mean('train_loss', dtype=tf.float32)
train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy('train_accuracy')

# define a single training step
@tf.function(input_signature=train_step_signature)
def train_step(inp, tar, hidden):
    with tf.GradientTape() as tape:
        predictions, hidden = generator(inp,hidden)

        loss = loss_function(tar, predictions)

    gradients = tape.gradient(loss, generator.trainable_variables)

    optimizer.apply_gradients(zip(gradients, generator.trainable_variables))

    train_loss(loss)
    train_accuracy(tar, predictions)

for epoch in range(epochs):
    start = time.time()
    
    # initializing the hidden state at the start of every epoch
    hidden = generator.initialize_hidden_state()
    
    # loop through training set
    steps_per_epoch = 0
    for (batch, (inp, tar)) in enumerate(dataset_batched_padded):
        train_step(inp, tar, hidden)

        if batch % 1 == 0:
            sys.stdout.write("Epoch %d Batch %d Loss %.4f \r" % (
                epoch + 1, batch, train_loss.result()
            ) )
            sys.stdout.flush()
        
        steps_per_epoch += 1
            
    print('-' * 89)
    print ('Epoch {} Train Loss {:.4f} Train Accuracy {:.1f} Time {:.2f}'.format(epoch + 1, 
                                                    train_loss.result(),
                                                    train_accuracy.result()*100,
                                                    (time.time() - start)))

    # Reset metrics every epoch
    train_loss.reset_states()
    train_accuracy.reset_states()
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it below
 
- with [CPU](https://colab.research.google.com/gist/amahendrakar/26baf21ddc0a48520870e5984b4437d9/40805-cpu.ipynb)

- with [GPU](https://colab.research.google.com/gist/amahendrakar/346729476adc2d38e0b11feb66af1854/40805-gpu.ipynb)

Thanks!
